snuba/clickhouse/formatter/expression.py

```diff
@@ -57,26 +53,39 @@ def __alias(self, formatted_exp: str, alias: Optional[str]) -> str:
             self.__parsing_context.add_alias(alias)
             return f"({formatted_exp} AS {escape_alias(alias)})"
 
+    @abstractmethod
+    def _format_string_literal(self, exp: Literal) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _format_number_literal(self, exp: Literal) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _format_boolean_literal(self, exp: Literal) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _format_datetime_literal(self, exp: Literal) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _format_date_literal(self, exp: Literal) -> str:
+        raise NotImplementedError
+
     def visit_literal(self, exp: Literal) -> str:
```

Comment:
hmm since we are modifying basically all of this function now I wonder if it's necessary to split out the methods for each type

Reply:
See my concern on the point above. By the end of this week we are likely to customize a lot more than these two methods to be smarter in what to redact.

In general though, the principle of the template method does not become more or less valid whether we are customizing a lot or not with respect to the base class. The issue it addresses is to clearly define the contract between what subclasses are allowed (and required) to extend vs what they are not allowed to extend. Knowing this contract allows developers to know what they can safely change in the base class vs what they have to verify on all subclasses. It remains valid whether you extend 1 or 10 methods.